### PR TITLE
Trim agent binary by 14mb by conditionally compiled XDS filters

### PR DESCRIPTION
**Pull request by howardjohn**
Part of https://github.com/istio/istio/issues/26232

75mb -> 64mb

We could have also done massive refactoring to avoid using build tags,
but I think in the near future we will need build tags for dropping k8s
import, so it seems useful to implement this at least for the short
term.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by mandarjog** [comment]
Looks good, I am verifying.

**Comment by howardjohn** [comment]
btw, forgot to mention, the reason we need the imports for pilot but not agent is pilot will marshal/unmarshal arbitrary Any messages for envoyfilter. Agent only unmarshals the top level DiscoveryReq/Resp

**Comment by kyessenov** [comment]
We should probably make sure unit tests for agent also have the build flag.

**Review by mandarjog — APPROVED** [review]
Lgtm sans the unit test comment.

**Review by ramaraochavali — APPROVED** [review]

File: Makefile.core.mk [hunk 1]
```diff
@@ -262,10 +262,15 @@ fmt: format-go format-python tidy-go
 buildcache:
 	GOBUILDFLAGS=-i $(MAKE) -e -f Makefile.core.mk build
 
+# gobuild script uses custom linker flag to set the variables.
```
Comment (mandarjog) [inline Makefile.core.mk:265]: https://github.com/istio/istio/blob/2a4a3889292bb9b8893ec69e352b8d4d2a5e329e/common/scripts/gobuild.sh#L46

I would just update the default ldflags here. stripping debug info should be the default.

**Pull request merged**